core/src/main/java/org/apache/iceberg/ManifestReader.java

```diff
@@ -78,11 +80,27 @@ public static ManifestReader read(InputFile file) {
    * @param specsById a Map from spec ID to partition spec
    * @return a manifest reader
    */
+  // TODO: do we want to keep it?
   public static ManifestReader read(InputFile file, Map<Integer, PartitionSpec> specsById) {
```

Comment:
Keeping this might be dangerous.

Reply:
The remaining usages are only in tests and `specsById` is always `null`. I am inclined to remove this method completely.

Reply:
What about people that are currently using this method? While we don't have a hard requirement to keep API compatibility before we release 1.0.0, I don't want to break people that are currently using the best practice. At Netflix, we have people that are using the version above as well.

I would rather keep this in and create a version of `InheritableMetadata` that fails if any metadata is missing. That way, we can keep these methods but safely guarantee that metadata is always present.

Reply:
That makes sense. I updated the existing methods to fail if they are used to read manifests with partial metadata and added new methods to read entries as they are without inheriting snapshot ids.